lib/spack/spack/environment.py

```diff
@@ -1165,21 +1165,25 @@ def install(self, user_spec, concrete_spec=None, **install_args):
         self._install(concrete, **install_args)
 
     def _install(self, spec, **install_args):
-        # "spec" must be concrete
-        spec.package.do_install(**install_args)
-
-        if not spec.external:
-            # Make sure log directory exists
-            log_path = self.log_path
-            fs.mkdirp(log_path)
-
-            with fs.working_dir(self.path):
-                # Link the resulting log file into logs dir
-                build_log_link = os.path.join(
-                    log_path, '%s-%s.log' % (spec.name, spec.dag_hash(7)))
-                if os.path.lexists(build_log_link):
-                    os.remove(build_log_link)
-                os.symlink(spec.package.build_log_path, build_log_link)
+        try:
+            # "spec" must be concrete
+            spec.package.do_install(**install_args)
+
+            if not spec.external:
+                # Make sure log directory exists
+                log_path = self.log_path
+                fs.mkdirp(log_path)
+
+                with fs.working_dir(self.path):
+                    # Link the resulting log file into logs dir
+                    build_log_link = os.path.join(
+                        log_path, '%s-%s.log' % (spec.name, spec.dag_hash(7)))
+                    if os.path.lexists(build_log_link):
+                        os.remove(build_log_link)
+                    os.symlink(spec.package.build_log_path, build_log_link)
+        except (Exception, SystemExit) as exc:
```

Comment:
@tldahlgren this comment appears to have been edited strangely: is this the end result you intended given the starting state of the comment?

Reply:
My response at the time was only to that part of the second bullet.  I keep forgetting to find and respond to the original bullet points.

Reply:
Uh oh.  I now see the issue.  I accidentally edited your original comment.

Reply:
> Two requests:
> 
>     * instead of using the try/catch here, can you use it in install_all? I think that since the entire function body here ends up being put into the try/catch, it would be simpler to guard the call to _install that occurs in install_all.

I suspect you are correct.  Let me look at that again.


> 
>     * Which process ends up raising SystemExit? Is something in do_install calling tty.die? I think we should avoid raising that exception (and catching it). 

IIRC It was the installation of the slate package that was raising `SystemExit`, which precluded lock cleanup.


>Also, is it possible to catch a more-specific exception than Exception?

I'll have to investigate this further.

Reply:
> > Two requests:
> > ```
> > * instead of using the try/catch here, can you use it in install_all? I think that since the entire function body here ends up being put into the try/catch, it would be simpler to guard the call to _install that occurs in install_all.
> > ```
> 
> I suspect you are correct. Let me look at that again.

I added the `try`/`catch` block in `_install` because `_install` is called from `install` and `install_all`.  @scheibelp Are you suggesting I move the block to both calls?

Reply:
> 
> > Also, is it possible to catch a more-specific exception than Exception?
> 
> I'll have to investigate this further.

I used `Exception` here since it has been used by the package install process.

Reply:
> I added the try/catch block in _install because _install is called from install and install_all. @scheibelp Are you suggesting I move the block to both calls?

It would only need to be caught in `install_all` though, wouldn't it? The problem was the installation of multiple root specs. If it in fact would need to be caught in both functions, then I agree that it's best to catch the exception here.

> I used Exception here since it has been used by the package install process.

As in there is code there which raises a generic `Exception`? I'd assume most parts of Spack raise `SpackError`.

Also, if there are packages which raise a `SystemExit`, then I'm inclined to say we should allow it to terminate the program. If people want different behavior, they should throw a different exception.